### ISSUE TITLE
Support for Array as column value

### ISSUE THREAD

**Issue by jcristovao** [vitaly-t/pg-promise#13]
Hi again,

It is stated in the readme that:

> pg-promise automatically converts all basic javascript types
> (text, boolean, date, number and null) into their Postgres format.

Is there any possibility of also supporting arrays, declared in postgres like:

```xpto int[],```

Currently trying to pass an array gives a:

> Cannot convert type 'object' of parameter with index ...

Thanks!

**Comment by vitaly-t** [comment]
How would you expect an array - parameter to be resolved into query values?

Please give us a full example, so we can understand better what it is that you are expecting.

**Comment by vitaly-t** [comment]
Ok, I have looked at the arrays in Postgres and I can see they must go in using the syntax of:

* `[1,2,3]` becomes `'{1,2,3}'`
* [[1,2],[3,4]] becomes `'{{1,2},{3,4}}'`

**Comment by jcristovao** [comment]
Exactly! That's it.
Just to be clear, I'm talking about assigning an array to a single ```$1``` value.

**Comment by vitaly-t** [comment]
>Just to be clear, I'm talking about assigning an array to a single $1 value.

Here's the first problem: When you pass an array to the formatting function it thinks you are using `$1, $2,..` formatting, so it cannot tell this from when you are passing it as a field value. This means you could only pass it either as `[[1,2,3]]` or as `{myArray: [1,2,3]}`

**Comment by jcristovao** [comment]
>  you should be able to do an easy work-around by converting such an array into a text string yourself.

Ok, fair enough, I shall give it a try! Thanks!

**Comment by vitaly-t** [comment]
I've just done the conversion helper, to support arrays of any dimension:

```javascript
// Converts array-value into Postgres format;
// It supports arrays of any dimension;
function formatArray(arr) {
    var s = '{';
    for (var i = 0; i < arr.length; i++) {
        if (i) {
            s += ',';
        }
        if (arr[i] instanceof Array) {
            s += formatArray(arr[i]);
        } else {
            s += arr[i]; // or converted;
        }
    }
    return s + '}';
}
````

I suppose there is no point limiting the number of dimensions supported.

One important aspect of such support: It is impossible to control that the array values adhere to Postgres support for arrays, i.e. we can only convert the values exactly as they are in the array, which means if you have an array of `[1, 'two', true]`, you get `'{1,'two',true}'`, which will never work with Postgres, so you have to control it yourself what's in array.

Do you think this would present a problem? For the time being I do not see what can we do anyway.

**Comment by jcristovao** [comment]
No, I think that's perfectly fine - one must know the limitations of postgres arrays.

Thanks!

**Comment by vitaly-t** [comment]
Ok, I have checked in the implementation + a few basic tests, it seems to be working fine now.

Method `pgp.as.array()` has been added also.

I will need to add full test coverage + update the documentation before I can release it, which will be either later today or tomorrow.

In the meantime, if you want to give it a go - just override your local formatting.js file with the latest from the master.

Let me know how this goes ;)

**Comment by vitaly-t** [comment]
UPDATE: The formatting method has been updated to correctly throw errors when it cannot convert the array:

```javascript
// Converts array-value into Postgres format;
// It supports arrays of any dimension;
function formatArray(arr, dim, step) {
    if (!dim) {
        dim = 0; // current array dimension;
    }
    if (!step) {
        step = 1; // current iteration step;
    }
    var s = '{';
    for (var i = 0; i < arr.length; i++) {
        if (i) {
            s += ',';
        }
        var v = arr[i];
        if (v instanceof Array) {
            s += formatArray(v, ++dim, step);
        } else {
            var value = formatValue(v);
            if (value === null) {
                throw new Error("Cannot convert type '" + typeof(v) + "' in array: level " + dim + ", index " + i + ", step " + step);
            }
            s += value;
        }
        step++;
    }
    return s + '}';
}
```

When fails, it will report the dimension/level + index within that dimension + the total interation step. I will document it later on.

**Comment by vitaly-t** [comment]
This one is much better, it reports the index properly in case of error...

```javascript
// Converts array-value into Postgres format;
// It supports arrays of any dimension;
function formatArray(arr, idx) {
    if (!idx) {
        idx = []; // indexes;
    }
    var s = '{';
    for (var i = 0; i < arr.length; i++) {
        if (i) {
            s += ',';
        }
        var v = arr[i];
        if (v instanceof Array) {
            idx.push(i);
            s += formatArray(v, idx);
        } else {
            var value = formatValue(v);
            if (value === null) {
                idx.push(i);
                throw new Error("Cannot convert type '" + typeof(v) + "' of array element with index " + idx.join(','));
            }
            s += value;
        }
    }
    return s + '}';
}
```

**Comment by vitaly-t** [comment]
Sorry, couldn't wait for you to get back on this, the feature has been fully tested and released as 0.9.2.

Thank you for the idea.

**Comment by vitaly-t** [comment]
Reopening this issue, because the initial implementation proved to work only for arrays of integers. Once you try using it for arrays of strings, nothing works there because of the way strings are formatted as array values - they use double quotes instead of single quotes.

In order to fix this problem a change has been made to use the alternative syntax of `array[]` supported by Postgres. This syntax allows any type of values in an open format (not as a text string), and text strings are passed using single quotes, so no issue there any more.

The fix has been committed and will be made available with 0.9.3 release.

**Comment by vitaly-t** [comment]
The last fix was released with 0.9.3

**Comment by chrisvariety** [comment]
Are you sure the alternate syntax `array[]` works with strings?

Quick POC:

```
pgp = require('pg-promise')();  
db = pgp('postgresql://localhost/whatever'); 

  db.none('create temporary table foobar ( test text[])').then(function() { 
  return db.none('insert into foobar values ($1)', pgp.as.array(['biz'])); 
}).then(function() {
  console.log('done', arguments);
}).catch(function() { 
  console.log('blew up', arguments); 
});
```

For me, this logs:

[error: malformed array literal: "array['biz']"] Array value must start with "{" or dimension information.

pg-promise 1.4.0, psql 9.4.1

according to [this comment](https://github.com/dockyard/postgres_ext/issues/31#issuecomment-9883844) PostgreSQL will not accept ARRAY[] syntax in a prepared insert statement.

**Comment by vitaly-t** [comment]
There is a formatting mistake in your example: First, you convert your array into a text string and then you inject array as a text string, which is, of course, incorrect.

Correct alternatives are shown below:

* regular array insertion:
```javascript
return db.none('insert into foobar values ($1)', [['biz']]));
```

* inserting array with explicit conversion:
```javascript
return db.none('insert into foobar values ($1^)', pgp.as.array(['biz']));
```
- note in the last one, we inject it as a raw text, not formatted text.

You can also insert it using the Named Parameters. [See all the examples](https://github.com/vitaly-t/pg-promise/wiki/Learn-by-Example#passing-arrays).

**Comment by chrisvariety** [comment]
@vitaly-t Got it! thanks so much for your quick reply--

**Comment by vitaly-t** [comment]
@chrisvariety , you are welcome!

And as for the link that you added, that is quite interesting. I wasn't aware of an issue like that, but then again, it was logged back in 2012, and PostgreSQL had a few releases since. The oldest I tested it against was 9.2, and it didn't have any problem accepting `array[]` syntax with inserts. I have to presume this issue was resolved in one of the earlier versions of PostgreSQL. And for that reason, I don't think there is any point in looking back at the old `{}` format, for it is quite old and awkward when it comes to dealing with arrays of strings.
